codec/src/main/java/io/netty/handler/codec/compression/ZstdDecoder.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.compression;
+
+import com.github.luben.zstd.Zstd;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+
+import static io.netty.handler.codec.compression.ZstdConstants.*;
+
+/**
+ * Decompresses a {@link ByteBuf} using the Zstandard algorithm.
+ * See <a href="https://facebook.github.io/zstd">Zstandard</a>.
+ */
+public final class ZstdDecoder extends ByteToMessageDecoder {
+
+    /**
+     * Current state of stream.
+     */
+    private enum State {
+        DECOMPRESS_DATA,
+        FINISHED,
+        CORRUPTED
+    }
+
+    private State currentState = State.DECOMPRESS_DATA;
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        try {
+            while (in.isReadable()) {
+                switch (currentState) {
+                    case DECOMPRESS_DATA:
+                        ByteBuf uncompressed = null;
+                        int decompressedLength = in.readableBytes();
+                        try {
+                            ByteBuffer src =  CompressionUtil.safeNioBuffer(in, in.readerIndex(), decompressedLength);
+                            int bufferSize = Math.max(decompressedLength, BUFFER_SIZE);
+                            uncompressed = ctx.alloc().directBuffer(bufferSize);
+                            ByteBuffer dst = uncompressed.internalNioBuffer(0, bufferSize);
+                            int srcSize = Zstd.decompress(dst, src);
```

Comment:
`Zstd.decompress` has overloads that work with byte arrays. Depending on the memory type of `in` ByteBuf, consider using an appropriate method.